### ISSUE TITLE
Cleanup: Reduce duplication of fire code

### ISSUE THREAD

**Issue by samsrabin** [ESCOMP/CTSM#2578]
There's substantial duplication among [CNFireLi2014Mod](https://github.com/ESCOMP/CTSM/blob/master/src/biogeochem/CNFireLi2014Mod.F90), [CNFireLi2016Mod](https://github.com/ESCOMP/CTSM/blob/master/src/biogeochem/CNFireLi2016Mod.F90), and [CNFireLi2021Mod](https://github.com/ESCOMP/CTSM/blob/master/src/biogeochem/CNFireLi2021Mod.F90). Once @lifang0209 finishes her current work, let's combine these into one file with conditionals based on the `fire_method` namelist parameter.

**Comment by ekluzek** [comment]
I do highly endorse making changes to reduce code duplication. However, I think I disagree with the suggested change in implementation, but we should talk about it. I think the overall OO design here is helpful, with the main problem being that the subroutines are too big, so the extensions end up duplicating earlier versions.

What I think would be better would be to have smaller methods with the base class, and the follow on extensions so that the duplicated parts can be shared between all the extensions. This is something I've wanted to do for a long time for the fire code.

Conditionals that are scattered in the code make the logic more complex and the code hard to follow. We have that throughout the code already and it would be good to reduce that pattern. OO design is one way to do this. In this case I think the OO code mainly needs to be examined in terms of how to split it up into smaller functions to reduce code duplication.

Anyway, @samsrabin this is something for you and I to discuss. I'd love to hear more about what you are thinking. And brainstorm on how to improve the current code. It's also exciting for me that it's realistic that you'll be able to work on this which would be awesome.

Also there are some considerations from long ago in #12 that we should maybe look at in this context. I thought there was an issue that I would have made around this, but I don't see it. And I don't see any reminders in the code that talk about it either.

**Comment by ekluzek** [comment]
Releasing this from the ctsm6.1.0 milestone, as that is going to be changed into the ctsm5.4 milestone for CMIP7 dataset updates.
